src/content/en/updates/2019/06/image-support-for-async-clipboard.md

```diff
@@ -0,0 +1,186 @@
+    console.log('Page URL copied to clipboard');
+  } catch (err) {
+    console.error('Failed to copy: ', err);
+  }
+};
+```
+
+### Paste: Reading text from the clipboard
+
+Much like copy, text can be read from the clipboard by calling `readText()`
+and waiting for the returned Promise to resolve with the text:
+
+```js
+const getClipboardContents = async () => {
+  try {
+    const text = await navigator.clipboard.readText();
+    console.log('Pasted content: ', text);
+  } catch (err) {
+    console.error('Failed to read clipboard contents: ', err);
+  }
+};
+```
+
+### Handling paste events
+
+Paste events can be handled by using the, surprise, `"paste"` event.
+It works nicely with the new asynchronous methods for reading clipboard text:
+
+```js
+document.addEventListener('paste', async (e) => {
+  e.preventDefault();
+  const text = await navigator.clipboard.readText()
+  console.log('Pasted text: ', text);
+});
+```
+
+### Security and permissions
+
+As usual, `navigator.clipboard` is only supported for pages served over HTTPS
+and to help prevent abuse, clipboard access is only allowed when a page is the active tab.
+Pages in active tabs can write to the clipboard without requesting permission,
+but reading from the clipboard always requires permission.
+
+Two new permissions for copy and paste were added to the
+[Permissions API](/web/updates/2015/04/permissions-api-for-the-web).
+The `clipboard-write permission` is granted automatically to pages when they are the active tab.
+The `clipboard-read permission` must be requested, which you can do by trying to read data from the clipboard.
+Attempting to read or write clipboard data will automatically prompt the user for permission
+ if it hasn not already been granted.
+
+```js
+const permissionStatus = await navigator.permissions.query({
+  name: 'clipboard-read'
+});
+// Will be 'granted', 'denied' or 'prompt':
+console.log(permissionStatus.state);
+
+// Listen for changes to the permission state
+permissionStatus.onchange = () => {
+  console.log(permissionStatus.state);
+};
+```
+
+## The new image-focused portion of the Asynchronous Clipboard API
+
+### Copy: Writing an image to the clipboard
+
+```js
+try {
+  await navigator.permissions.request({name: 'clipboard-write'});
+  const data = await fetch('https://upload.wikimedia.org/wikipedia/commons/5/53/Google_Chrome_Material_Icon-450x450.png');
+  const blob = await data.blob();
+  await navigator.clipboard.write([
+    new ClipboardItem(Object.defineProperty({}, blob.type, {
+      value: blob,
+      enumerable: true
+    }))
+  ]);
+  console.log('Image copied.');
+} catch(e) {
+  console.error(e, e.message);
+}
+```
+
+### Paste: Reading an image from the clipboard
+
+```js
+try {
+  await navigator.permissions.request({name: 'clipboard-read'});
+  const clipboardItems = await navigator.clipboard.read();
+  clipboardItems.forEach((clipboardItem) => {
+    try {
+      clipboardItem.types.forEach(async (type) => {
```

Comment:
I think that this will not do what you expect it to. Here you have an `Array.prototype.forEach` call to which you pass an async arrow function. `Array.prototype.forEach` will call the async arrow function for each  `clipboardItem`'s `type` and get back a `Promise` each time, but it will do nothing with it. As soon as it calls the async arrow function for each `clipboardItem`'s `type`, it will just return. So, we exit the `try catch` block even before any of those promises has the chance to resolve/reject. To fix this, you should use the `for...of` loop instead of `Array.prototype.forEach` both when iterating over `clipboardItems` and `clipboardItem`'s `type`'s. The `for...of` loop will do what you expect and it's also more readable.

Reply:
Duh, good catch! Thanks!